vast/opensetAlgos/WPL.py

```diff
@@ -0,0 +1,140 @@
+
+
+def WPL_Params(parser):
+    WPL_params = parser.add_argument_group("WPL params")
+    WPL_params.add_argument(
+        "--tailsize",
+        nargs="+",
+        type=float,
+        default=[1.0],
+        help="tail size to use default: %(default)s",
+    )
+    WPL_params.add_argument(
+        "--distance_multiplier",
+        nargs="+",
+        type=float,
+        default=[1.0],
+        help="distance multiplier to use default: %(default)s",
+    )
+    WPL_params.add_argument(
+        "--default_scale",
+        nargs="+",
+        type=float,
+        default=[1.0],
+        help="Weibull scale to use when the number of uniqe element is less than 5 default: %(default)s",
+    )
+    WPL_params.add_argument(
+        "--default_shape",
+        nargs="+",
+        type=float,
+        default=[0.1],
+        help="Weibull shape to use when the number of uniqe element is less than 5 default:  %(default)s",
+    )
+    return parser, dict(
+        group_parser=WPL_params,
+        param_names=("tailsize", "distance_multiplier", "default_scale", "default_shape"),
+        param_id_string="TS_{}_DM_{:.4f}_SC_{:.4f}_SH_{:.4f}",
+    )
+
+
+def fit_high(distances, distance_multiplier, tailsize, default_shape, default_scale):
+    distances = torch.unique(distances)
+    if tailsize <= 1:
+        tailsize = min(tailsize * distances.shape[1], distances.shape[1])
+    tailsize = int(min(tailsize, distances.shape[1]))
+    mr = weibull.weibull()
+    if distances.shape[1] < 5:
+        pass
+    else:
+        mr.FitHigh(distances.double() * distance_multiplier, tailsize, isSorted=False)
+    mr.tocpu()
+    return mr
+
+  
+def WPL_Training(
+    pos_classes_to_process: List[str],
+    features_all_classes: OrderedDict[str, torch.Tensor],
+    args,
+    gpu: int,
+    models=None,
+) -> Iterator[Tuple[str, Tuple[str, dict]]]:
+    """
+    :param pos_classes_to_process: List of class names to be processed by this function in the current process class.
+    :param features_all_classes: features of all classes, note the classes in pos_classes_to_process can be a subset of the keys for this ordered dictionary
+    :param args: A named tuple or an argument parser object containing the arguments mentioned in the WPL_Params function above.
+    :param gpu: An integer corresponding to the gpu number to use by the current process.
+    :param models: Not used during training, input ignored.
+    :return: Iterator(Tuple(parameter combination identifier, Tuple(class name, its evm model)))
+    """
+    with torch.no_grad():
+        for pos_cls_name in pos_classes_to_process:
+            features = features_all_classes[pos_cls_name].clone().to(f"cuda:{gpu}")
+            assert args.dimension == features.shape[1]
+
+            center = torch.mean(features, dim=0).to(f"cuda:{gpu}")
+            distances = torch.abs(features - center.view(1,args.dimension).repeat(features.shape[0], 1))
```

Comment:
yes, you are right that this might not be replaceable by other metrics and neither would be a useful metric for other algorithms, which is what I was thinking.
BTW, you do not need `.repeat` above, even if needed it should be replaced with `expand` as it reduces memory consumption.